File: flink-train/src/main/scala/com/tqz/scala/course05/DataStreamDataSourceApp.scala
package com.tqz.scala.course05

import org.apache.flink.api.scala._
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.streaming.api.windowing.time.Time

/**
  * 使用 Scala 做数据流Demo
  *
  * 注意批处理的执行环境是 ExecutionEnvironment
  * 但是流处理的执行环境是 StreamExecutionEnvironment，并且流处理最后要执行 execute 方法
  */
object DataStreamDataSourceApp {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    //socketFunction(env)
//    customNonParallelSource(env)
//    customParallelSource(env)
    customRichParallelSource(env)
    env.execute("Window Stream WordCount")
  }

  def customRichParallelSource(env: StreamExecutionEnvironment): Unit = {
    // 可以并行执行
    val text = env.addSource(CustomRichParallelSourceFunction).setParallelism(2)
    text.print()
  }

  def customParallelSource(env: StreamExecutionEnvironment): Unit = {
    // 可以并行执行
    val text = env.addSource(CustomParallelSourceFunction).setParallelism(2)
    text.print().setParallelism(1)
  }

  def customNonParallelSource(env: StreamExecutionEnvironment): Unit = {
    // 不可以并行执行
    val text = env.addSource(CustomNonParallelSourcesFunction).setParallelism(1) // 只能是 1
    text.print()
  }

  def socketFunction(env: StreamExecutionEnvironment): Unit = {
    val text = env.socketTextStream("54bruce.com", 8580)

    val counts = text.flatMap {
      _.toLowerCase.split(",") filter {
        _.nonEmpty
      }
    }
      .map {
        (_, 1)
      }
      .keyBy(0)
      .timeWindow(Time.seconds(5))
      .sum(1)

    counts.print()
  }

}

File: src/main/java/com/tqz/java/course05/DataStreamDataSourceApp.java
package com.tqz.java.course05;

import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.util.Collector;


import java.util.Arrays;

/**
 * 注意批处理的执行环境是 ExecutionEnvironment
 * 但是流处理的执行环境是 StreamExecutionEnvironment，并且流处理最后要执行 execute 方法
 */
public class DataStreamDataSourceApp {

    public static void main(String[] args) throws Exception{
        // 1. 获取执行环境
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        // 2. 获取数据  & 3. 操作数据
        socketFunction(env);

        // 4. 执行程序
        env.execute("DataStreamDataSourceApp");
    }

    public static void socketFunction(StreamExecutionEnvironment env) {
        // 2. 获取数据
        DataStreamSource<String> text = env.socketTextStream("120.79.241.167", 8580);

        // 3. 操作数据：统计单词频率
        text.flatMap(new FlatMapFunction<String, Tuple2<String, Integer>>() {
            @Override
            public void flatMap(String in, Collector<Tuple2<String, Integer>> out) throws Exception {
                String[] tokens = in.split(",");
                for (String token: tokens) {
                    if (token.length() > 0) {
                        out.collect(new Tuple2<>(token, 1));
                    }
                }
            }
        }).keyBy(0)
                //.countWindow(1) // 每个窗口里面只要放入了一个元素就会被触发
                .countWindow(2) // 每个窗口里面只要放入了 2 个元素就会被触发
                .sum(1).print();

    }
}
